### PR TITLE
attempt to fix relative pathing error

### PR DESCRIPTION
**Pull request by valarauca**
I threw this together in an attempt to address: https://github.com/softdevteam/grmtools/issues/185

I'm not sure if you want to fix it in house, or use an external crate.

### CODE REVIEW

**Comment by ltratt** [comment]
I agree that #185 is a problem that needs fixing, so thanks for looking into this! I don't know whether we should fix this in-house or use a crate. A quick look suggested that something like https://docs.rs/include_dir/0.6.0/include_dir/ might solve things for us? Then my question would probably be whether we think our solution is stable going forward (at which point not dragging in another dependency is probably a good thing) or whether we want to allow some other poor soul to keep things working in the future (at which point a dependency is probably worth the cost). Any thoughts?

**Comment by valarauca** [comment]
> Any thoughts?

The whole situation could also be avoided by just serializing into a `const NAME: &'static [u8]` instead of other files.

Generated sources almost always lead to "_ugly diffs_" in git. I doubt this would surprise/annoy down stream users. As an example [`grpc`/`protobuf`](https://github.com/grpc/grpc-go/blob/master/examples/route_guide/routeguide/route_guide.pb.go#L315) include the base source used to generate the file, so it can communicate metadata.

**Comment by ltratt** [comment]
> The whole situation could also be avoided by just serializing into a const NAME: &'static [u8] instead of other files.

I'm not sure I understand -- can you expand a little?

**Comment by ltratt** [comment]
> The whole situation could also be avoided by just serializing into a const NAME: &'static [u8] instead of other files.

Oh, I have just realised what you mean!

Are there any downsides to doing this? If not, maybe that's the most stable long-term fix? I guess we would need to check that it doesn't tank compile times though.

**Comment by ltratt** [comment]
[The particular mechanism we use is intentionally hidden from the user as the idea "one day" (in the mystical perfect future) is that we generate much better code for the statetable. So provided none of our changes leak out to the user, I don't have a strong opinion on how we do things -- `include_bytes` was just the first thing that came to my mind when I implemented this way back when.]

**Comment by valarauca** [comment]
Instead of serializing the data to disk, as a file, and loading that file via `include_bytes!()`.

The data could just be serialized to a Rust-Array within the generated source code directly.

**Comment by ltratt** [comment]
> The data could just be serialized to a Rust-Array within the generated source code directly.

Yes, I was slow to realise what you meant.

If this doesn't tank compile times, I think it might be the neatest solution? It's certainly the least code to maintain.

**Comment by ltratt** [comment]
Hello @valarauca, I realise my last message probably wasn't very clear. Basically I'm open-minded about which solution you think is best (the one you've initially done in the PR or the "embed everything in the string"), and would welcome your opinion about which you think is the best long-term fix. Any thoughts?

**Comment by valarauca** [comment]
I think the best long term solution would be the constant.

**Comment by ltratt** [comment]
Would you like to do that for this PR, or shall we merge this then do the constant?

**Comment by valarauca** [comment]
I'll take a crack at this afternoon.

**Comment by valarauca** [comment]
2 new commits.

`inline_constants` is the bulk of the work. It does indeed convert the values into inline constants.

The second commit modifies the formatting. I was originally using hexadecimal for the generated byte arrays, but decimal seems to save a fair amount of space.

On a local project, with a fairly trivial grammar. Pre-Patch the generated file was `19KiB`, Post-Patch it is `30KiB`. Using Hexadecimal the generated file size was `40KiB`, that is why I patched in decimal support. 

---

I haven't seen a change in compile times.

File: lrpar/src/lib/ctbuilder.rs [hunk 1]
```diff
@@ -492,22 +492,13 @@ where
...
         };
 
-        // Because we're lazy, we don't write our own serializer. We use serde and bincode to
-        // create files $out_base.grm, $out_base.sgraph, and $out_base.out_stable which contain
-        // binary versions of the relevant structs, and then include those binary files into the
-        // Rust binary using "include_bytes" (see below).
-        let out_grm = self.bin_output(&outp_base, GRM_FILE_EXT, &grm)?;
-        let out_sgraph = self.bin_output(&outp_base, SGRAPH_FILE_EXT, &sgraph)?;
-        let out_stable = self.bin_output(&outp_base, STABLE_FILE_EXT, &stable)?;
+        // bincode constants are deserialized within
```
Comment (ltratt) [inline lrpar/src/lib/ctbuilder.rs:495]: This comment seems a bit unfinished?

File: lrpar/src/lib/ctbuilder.rs [hunk 2]
```diff
@@ -892,57 +869,57 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
...
+        Self {
+            buffer: format!(r#"#[allow(dead_code)] const {}: &[u8] = &["#, name),
         }
-        Option::None => {
-            // nothing in the paths is shared.
-            PathBuf::from(other)
+    }
+
+    /// complete the array, and return the finished string
+    fn finish(self) -> String {
```
Comment (ltratt) [inline lrpar/src/lib/ctbuilder.rs:897]: I *think* (but I might be wrong) we could mildly simplify this to:

```rust
fn finish (mut self) {
    buffer.push_str(...);
    buffer
}
```

?

File: lrpar/src/lib/ctbuilder.rs [hunk 3]
```diff
@@ -892,57 +869,57 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
...
+        let mut buffer = self.buffer;
+        buffer.push_str("];\n");
+        buffer
+    }
+}
+impl Write for ArrayWriter {
+    #[allow(dead_code)]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        use std::fmt::Write;
```
Comment (ltratt) [inline lrpar/src/lib/ctbuilder.rs:906]: This should probably go with the other `use`s at the top of the file?

File: lrpar/src/lib/ctbuilder.rs [hunk 4]
```diff
@@ -906,7 +906,7 @@ impl Write for ArrayWriter {
         use std::fmt::Write;
         for b in buf {
             // specifying hexidecimal format.
```
Comment (ltratt) [inline lrpar/src/lib/ctbuilder.rs:908]: I think we can remove this comment?

**Comment by valarauca** [comment]
Good feedback. I'll get them addressed ASAP.

File: lrpar/src/lib/ctbuilder.rs (hunk 2 above)
Comment (valarauca) [reply lrpar/src/lib/ctbuilder.rs:897]: I think we'll still need the `self.` prefix. But I understand what you're saying.

Comment (ltratt) [reply lrpar/src/lib/ctbuilder.rs:897]: Mea culpa! Yes, that was silly of me.

File: lrpar/src/lib/ctbuilder.rs (hunk 1 above)
Comment (valarauca) [reply lrpar/src/lib/ctbuilder.rs:495]: I can just remove it.

Comment (ltratt) [reply lrpar/src/lib/ctbuilder.rs:495]: Works for me.

File: lrpar/src/lib/ctbuilder.rs (hunk 3 above)
Comment (valarauca) [reply lrpar/src/lib/ctbuilder.rs:904]: I don't think so. We already have `std::io::Write` within the namespace, so bringing in `std::fmt::Write` generates a collision.

Comment (ltratt) [reply lrpar/src/lib/ctbuilder.rs:904]: That's a good point -- I hadn't encountered that particular clash before. I *think* you can use `as` to rename the two things out of the way of each other. Whether that's worth doing or not is another thing.

Comment (valarauca) [reply lrpar/src/lib/ctbuilder.rs:904]: Hmm the `as` alias would work.

Comment (valarauca) [reply lrpar/src/lib/ctbuilder.rs]: Updated the PR to use it.

**Comment by ltratt** [comment]
This looks good to me! Can you `cargo fmt` (alas, using nightly `rustfmt` :/ )? Then you can squash the commits down as you see fit, and we can merge.

**Comment by ltratt** [comment]
Thanks!

bors r+

**Pull request merged**